### ISSUE TITLE
Use c-ares library to handle dns resolve

### ISSUE THREAD

**Issue by adiholden** [dragonflydb/dragonfly#947]
instead calling syncronious  function getaddrinfo
Use asynchronous calls from the following library - c-ares

**Comment by romange** [comment]
Dragonfly is fully asynchronous and uses fibers, while a regular DNS call like `getaddrinfo` is synchronous and does not fit Dragonfly/helio architecture. c-ares exposes a flexible API that allows resolving DNS queries asynchronously which should work well with helio. In fact the integration should be in helio because it also calls `getaddrinfo` in several places.

**Comment by chakaz** [comment]
I've prepared a skeleton of adding c-ares to helio: https://github.com/romange/helio/commit/7d39d07d30bb10a48ff8eaa608552e45bebef325

Some questions, as this is definitely not PR-worthy yet:
1. Am I using Fiber's futures correctly (specifically [here](https://github.com/romange/helio/commit/7d39d07d30bb10a48ff8eaa608552e45bebef325#diff-8da314b17125e2df7dab843471b6fda6461eba0175b93eb625d790f8cc18e4f4R67))? I want to make sure that the function is blocking, but is also yielding to other fibers using the thread, and I'm not sure if I got it right (first timer, sorry)
2. Should I replace the call also in accept_server.cc? ([here](https://github.com/romange/helio/blob/master/util/accept_server.cc#L95)) It looks like c-ares doesn't support `AI_PASSIVE`, and also maybe that implementation doesn't really block because it's a localhost? And also, adding listeners, hopefully, is not on the hot path anyway?
3. I couldn't figure if the call to `ares_init()` is expensive and I should cache it or not. If anyone knows, please lmk, otherwise I'll try to ask in some forum or whatnot

FWIW the way I tested it is via running an echo_server in server mode (actually without these changes), and then echo_server in client mode (via `--connect=127.0.0.1` and `--connect=localhost`) and it showed a valid report.

**Comment by romange** [comment]
I actually do not know an answer to any of your questions. It seems you use
futures correctly (have not used them myself a lot), regarding AI_PASSIVE
have no idea but if it works - it works :)

Also, it's super easy to create a HTTP client (see dfly_main.cc - we do it
for the version check).
if we do not have http_client in helio - maybe it's worth adding it to
examples. With it - you'll be able to test it vs the external domains (also
https, not sure if they require special handling).

**Comment by chakaz** [comment]
Gotcha, thanks! Can you kindly assign this task to me? I've had some progress with this feature-request (like adding it to helio, using it) that I'm confident enough that I can take it :)

(https://github.com/romange/helio/commit/95bae876926851b6e5b18fb7be2e7d5adb2fa037)

**Comment by chakaz** [comment]
I got some answers:
1. Channel should *not* be shared, it's a one-off, per request thing (it's not even thread safe)
2. There's already an HTTP client in helios, with a cool wrapper example (that does HTTPS): https_client_cli. I checked and my changes work with it indeed!
3. I did not modify the add listener part, as it's probably not worth touching (and risking), at least not for now.
Sending a PR now!
